File: Projects/Final/src/HomeHub/DataVisualization.js
google.charts.load('current', {packages: ['corechart', 'bar', 'table']});

$("#btnTableServices").click(function () {
	var x = document.getElementById("table_services");
	if(x.style.display == "none") {
		x.style.display = "block";
		document.getElementById("btnTableServices").innerHTML = "Close Table of Services";
		$.ajax({
			url: "DataVisualization",
			type: "POST",
			data: "action=table_services",
			success: function (msg) {
				createData("table_services", msg)
			},
			error: function(){
				console.log("error occurred while making ajax call")
			}
		});
	}else if(x.style.display == "block"){
		x.style.display = "none";
		document.getElementById("btnTableServices").innerHTML = "View Table of Services";
	}
});

$("#btnChartServices").click(function () {
	var x = document.getElementById("chart_services");
	if(x.style.display == "none") {
		x.style.display = "block";
		document.getElementById("btnChartServices").innerHTML = "Close Chart of Services";
		$.ajax({
			url: "DataVisualization",
			type: "POST",
			data: "action=chart_services",
			success: function (msg) {
				createData("chart_services", msg)
			},
			error: function(){
				console.log("error occurred while making ajax call")
			}
		});
	}else if(x.style.display == "block"){
		x.style.display = "none";
		document.getElementById("btnChartServices").innerHTML = "View Chart of Services";
	}
});

$("#btnTableServiceSold").click(function () {
	var x = document.getElementById("table_service_sold");
	if(x.style.display == "none") {
		x.style.display = "block";
		document.getElementById("btnTableServiceSold").innerHTML = "Close Table of Service Sold";
		$.ajax({
			url: "DataVisualization",
			type: "POST",
			data: "action=table_service_sold",
			success: function (msg) {
				createData("table_service_sold", msg)
			},
			error: function(){
				console.log("error occurred while making ajax call")
			}
		});
	}else if(x.style.display == "block"){
		x.style.display = "none";
		document.getElementById("btnTableServiceSold").innerHTML = "View Table of Service Sold";
	}
});

$("#btnChartServiceSold").click(function () {
	var x = document.getElementById("chart_service_Sold");
	if(x.style.display == "none") {
		x.style.display = "block";
		document.getElementById("btnChartServiceSold").innerHTML = "Close Chart of Service Sold";
		$.ajax({
			url: "DataVisualization",
			type: "POST",
			data: "action=chart_service_Sold",
			success: function (msg) {
				createData("chart_service_Sold", msg)
			},
			error: function(){
				console.log("error occurred while making ajax call")
			}
		});
	}else if(x.style.display == "block"){
		x.style.display = "none";
		document.getElementById("btnChartServiceSold").innerHTML = "View Chart of Service Sold";
	}
});

$("#btnTableDailySalesTransaction").click(function () {
	var x = document.getElementById("table_daily_sales_transaction");
	if(x.style.display == "none") {
		x.style.display = "block";
		document.getElementById("btnTableDailySalesTransaction").innerHTML = "Close Table of Daily Dales Transaction";
		$.ajax({
			url: "DataVisualization",
			type: "POST",
			data: "action=table_daily_sales_transaction",
			success: function (msg) {
				createData("table_daily_sales_transaction", msg)
			},
			error: function(){
				console.log("error occurred while making ajax call")
			}
		});
	}else if(x.style.display == "block"){
		x.style.display = "none";
		document.getElementById("btnTableDailySalesTransaction").innerHTML = "View Table of Daily Dales Transaction";
	}
});



function createData(id, jsonData) {

	if(id == "table_services"){
		var parsed_data = $.parseJSON(jsonData);
		var data_arr = new Array();
		for(var i=0; i<parsed_data.length; i++) {
			var cat_super = parsed_data[i]["category super"];
			var cat_sub = parsed_data[i]["category sub"];
			var service_count = parsed_data[i]["service count"];
			data_arr.push( [ cat_super , cat_sub , service_count ] );
		}
		drawTable(id, data_arr);
	}

	if(id == "chart_services"){
		var parsed_data = $.parseJSON(jsonData);
		var data_arr = new Array();
		for(var i=0; i<parsed_data.length; i++) {
			var cat_super = parsed_data[i]["category super"];
			var cat_sub = parsed_data[i]["category sub"];
			var service_count = parsed_data[i]["service count"];
			data_arr.push( [ "("+cat_super+","+cat_sub+")" , service_count ] );
		}
		drawChart(id, data_arr);
	}

	if(id == "table_service_sold"){
		var parsed_data = $.parseJSON(jsonData);
		var data_arr = new Array();
		for(var i=0; i<parsed_data.length; i++) {
			var service_name = parsed_data[i]["name"];
			var service_rate = parsed_data[i]["rate"];
			var service_hours = parsed_data[i]["hours"];
			var service_sales = parsed_data[i]["sales"];
			data_arr.push( [ service_name , service_rate , service_hours, service_sales ] );
		}
		drawTable(id, data_arr);
	}

	if(id == "chart_service_Sold"){
		var parsed_data = $.parseJSON(jsonData);
		var data_arr = new Array();
		for(var i=0; i<parsed_data.length; i++) {
			var service_name = parsed_data[i]["name"];
			var service_sales = parsed_data[i]["sales"];
			data_arr.push( [ service_name , service_sales ] );
		}
		drawChart(id, data_arr);
	}

	if(id == "table_daily_sales_transaction"){
		var parsed_data = $.parseJSON(jsonData);
		var data_arr = new Array();
		for(var i=0; i<parsed_data.length; i++) {
			var date = parsed_data[i]["date"];
			var sales = parsed_data[i]["sales"];
			data_arr.push( [ date , sales ] );
		}
		drawTable(id, data_arr);
	}

}

function drawTable(id, data){

	if(id == "table_services"){
		var table_data = new google.visualization.DataTable();
		table_data.addColumn('string', 'category super');
		table_data.addColumn('string', 'category sub');
		table_data.addColumn('number', 'service count');
		table_data.addRows(data);
		var options = {
			'width': 600,
			'height': 650,
			'showRowNumber': true,
		};
		var table = new google.visualization.Table(document.getElementById(id));
		table.draw(table_data, options);
	}

	if(id == "table_service_sold"){
		var table_data = new google.visualization.DataTable();
		table_data.addColumn('string', 'Service Name');
    table_data.addColumn('number', 'Service Rate');
		table_data.addColumn('number', 'Service Hours');
		table_data.addColumn('number', 'Service Sales');
		table_data.addRows(data);
		var options = {
			'width': 600,
			'height': 650,
			'showRowNumber': true,
		};
		var table = new google.visualization.Table(document.getElementById(id));
		table.draw(table_data, options);
	}

	if(id == "table_daily_sales_transaction"){
		var table_data = new google.visualization.DataTable();
		table_data.addColumn('string', 'Date');
    table_data.addColumn('number', 'Sales');
		table_data.addRows(data);
		var options = {
			'width': 600,
			'height': 650,
			'showRowNumber': true,
		};
		var table = new google.visualization.Table(document.getElementById(id));
		table.draw(table_data, options);
	}

}

function drawChart(id , data) {

	if(id == "chart_services"){
		data.unshift(["category-(super,sub)", "service count"]);
		var chart_data = google.visualization.arrayToDataTable(data);
		var options = {
			'width':800,
			'height':800,
			title: "Service Count Chart",
			bars: 'horizontal'
		};
		var chart = new google.visualization.BarChart(document.getElementById(id));
		chart.draw(chart_data, options);
	}

	if(id == "chart_service_Sold"){
		data.unshift(["Service Name", "Service Sales"]);
		var chart_data = google.visualization.arrayToDataTable(data);
		var options = {
			'width':800,
			'height':800,
			title: "Service Sales Chart",
			bars: 'horizontal'
		};
		var chart = new google.visualization.BarChart(document.getElementById(id));
		chart.draw(chart_data, options);
	}

}

File: Projects/Final/src/HomeHub/WEB-INF/classes/DataVisualization.java
import com.google.gson.Gson;
import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import com.mongodb.MongoClient;
import com.mongodb.DB;
import com.mongodb.DBCollection;
import com.mongodb.BasicDBObject;
import com.mongodb.DBObject;
import com.mongodb.DBCursor;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;
import com.mongodb.AggregationOutput;


@WebServlet("/DataVisualization")
public class DataVisualization extends HttpServlet {

  @Override
  protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
    response.setContentType("text/html");
    PrintWriter pw = response.getWriter();
    Utilities utility = new Utilities(request, pw);

    if(utility.isLoggedin() && utility.user_type().equals("administrator")) {
      displayReportPage(request, response, pw);
    }else{
      response.sendRedirect("Login");
      return;
    }
  }

  protected void displayReportPage(HttpServletRequest request, HttpServletResponse response, PrintWriter pw) throws ServletException, IOException {

    Utilities utility = new Utilities(request, pw);

    utility.printHtml("Head.html");
    utility.printAHeader();
    utility.printALeftNav();
		
	
		pw.write("<div class='w3-container w3-text-grey' id='all_content'>");
		pw.write("<h1 class='w3-hide-small' style='color:black'><center>Data Visualization</center></h1>");
		pw.write("</div>");

		pw.write("<div class='w3-container w3-text-grey' id='all_content'>");
		pw.write("<h2 class='w3-hide-small' style='color:black'>Inventory</h2>");
		pw.write("</div>");

		pw.print("<h3><button id='btnTableServices' class='btnreview w3-button w3-black'>View Table of Services</h3>");
		pw.println("<div id='table_services' style='display:none;'></div>");

		pw.print("<h3><button id='btnChartServices' class='btnreview w3-button w3-black'>View Chart of Services</h3>");
		pw.println("<div id='chart_services' style='display:none;'></div>");

		pw.write("<div class='w3-container w3-text-grey' id='all_content'>");
		pw.write("<h2 class='w3-hide-small'  style='color:black'>Sales Report</h2>");
		pw.write("</div>");

		pw.print("<h3><button id='btnTableServiceSold' class='btnreview w3-button w3-black'>View Table of Service Sold</h3>");
		pw.println("<div id='table_service_sold' style='display:none;'></div>");

		pw.print("<h3><button id='btnChartServiceSold' class='btnreview w3-button w3-black'>View Chart of Service Sold</h3>");
		pw.println("<div id='chart_service_Sold' style='display:none;'></div>");

		pw.print("<h3><button id='btnTableDailySalesTransaction' class='btnreview w3-button w3-black'>View Table of Daily Sales Transaction</h3>");
		pw.println("<div id='table_daily_sales_transaction' style='display:none;'></div>");

		pw.println("<script type='text/javascript' src=\"https://www.gstatic.com/charts/loader.js\"></script>");
		pw.println("<script type='text/javascript' src='DataVisualization.js'></script>");


		utility.printHtml("Footer.html");

  }

  @Override
  protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
    try {
			String action = request.getParameter("action");
			String dataJson = null;
      if(action.equalsIgnoreCase("table_services")){
        HashMap<Integer, Service> services = MySqlDataStoreUtilities.getServices();
        HashMap<String, HashMap<String, Integer>> map_services = new HashMap();

        for(Map.Entry<Integer, Service> entry : services.entrySet()){
          Service s = entry.getValue();
          if(s.service_status.equalsIgnoreCase("active")){
            String cat_super = s.service_category_super.toLowerCase();
            String cat_sub = s.service_category_sub.toLowerCase();

            HashMap<String, Integer> map_service = map_services.getOrDefault(cat_super, new HashMap());
            map_service.put(cat_sub, map_service.getOrDefault(cat_sub, 0) + 1);
            map_services.put(cat_super, map_service);
          }
        }

        ArrayList<HashMap<String, Object>> list_services = new ArrayList();
        for(Map.Entry<String,HashMap<String, Integer>> entry : map_services.entrySet()){
          for(Map.Entry<String, Integer> entry1 : entry.getValue().entrySet()){
            HashMap<String, Object> obj = new HashMap();
            obj.put("category super", entry.getKey());
            obj.put("category sub", entry1.getKey());
            obj.put("service count", entry1.getValue());
            list_services.add(obj);
          }
        }
        dataJson = new Gson().toJson(list_services);
      }
      if(action.equalsIgnoreCase("chart_services")){
        HashMap<Integer, Service> services = MySqlDataStoreUtilities.getServices();
        HashMap<String, HashMap<String, Integer>> map_services = new HashMap();

        for(Map.Entry<Integer, Service> entry : services.entrySet()){
          Service s = entry.getValue();
          if(s.service_status.equalsIgnoreCase("active")){
            String cat_super = s.service_category_super.toLowerCase();
            String cat_sub = s.service_category_sub.toLowerCase();

            HashMap<String, Integer> map_service = map_services.getOrDefault(cat_super, new HashMap());
            map_service.put(cat_sub, map_service.getOrDefault(cat_sub, 0) + 1);
            map_services.put(cat_super, map_service);
          }
        }

        ArrayList<HashMap<String, Object>> list_services = new ArrayList();
        for(Map.Entry<String,HashMap<String, Integer>> entry : map_services.entrySet()){
          for(Map.Entry<String, Integer> entry1 : entry.getValue().entrySet()){
            HashMap<String, Object> obj = new HashMap();
            obj.put("category super", entry.getKey());
            obj.put("category sub", entry1.getKey());
            obj.put("service count", entry1.getValue());
            list_services.add(obj);
          }
        }
        dataJson = new Gson().toJson(list_services);
      }
      if(action.equalsIgnoreCase("table_service_sold")){
				HashMap<Integer, Transaction> map_transaction = MySqlDataStoreUtilities.getTransactions();
				HashMap<String, HashMap<String, Object>> map_service_transactions = new HashMap();

				for(Map.Entry<Integer, Transaction> entry : map_transaction.entrySet()){
          Transaction t = entry.getValue();
          if(t.is_cancelled == false && t.service_status.equalsIgnoreCase("accepted") && t.transaction_status.equalsIgnoreCase("approved")){
            String service_name = t.service_name.toLowerCase();
            double service_rate = t.service_rate;
            int service_hours = t.service_hours;
            double service_sales = t.service_total_amount;

            HashMap<String, Object> map_service = map_service_transactions.getOrDefault(service_name, new HashMap());
            map_service.put("name", service_name);
            map_service.put("rate", service_rate);
            map_service.put("hours", (int)map_service.getOrDefault("hours", 0) + service_hours);
            map_service.put("sales", (double)map_service.getOrDefault("sales", 0.0) + service_sales);
            map_service_transactions.put(service_name, map_service);
          }
				}

				ArrayList<HashMap<String, Object>> list_service_transaction = new ArrayList();
				for(Map.Entry<String,HashMap<String, Object>> entry : map_service_transactions.entrySet()){
					list_service_transaction.add(entry.getValue());
				}
				dataJson = new Gson().toJson(list_service_transaction);
			}
			if(action.equalsIgnoreCase("chart_service_Sold")){
        HashMap<Integer, Transaction> map_transaction = MySqlDataStoreUtilities.getTransactions();
				HashMap<String, HashMap<String, Object>> map_service_transactions = new HashMap();

        for(Map.Entry<Integer, Transaction> entry : map_transaction.entrySet()){
          Transaction t = entry.getValue();
          if(t.is_cancelled == false && t.service_status.equalsIgnoreCase("accepted") && t.transaction_status.equalsIgnoreCase("approved")){
            String service_name = t.service_name.toLowerCase();
            double service_sales = t.service_total_amount;

            HashMap<String, Object> map_service = map_service_transactions.getOrDefault(service_name, new HashMap());
            map_service.put("name", service_name);
            map_service.put("sales", (double)map_service.getOrDefault("sales", 0.0) + service_sales);
            map_service_transactions.put(service_name, map_service);
          }
        }

        ArrayList<HashMap<String, Object>> list_service_transaction = new ArrayList();
        for(Map.Entry<String,HashMap<String, Object>> entry : map_service_transactions.entrySet()){
          list_service_transaction.add(entry.getValue());
        }
        dataJson = new Gson().toJson(list_service_transaction);
			}
			if(action.equalsIgnoreCase("table_daily_sales_transaction")){
				HashMap<Integer, Transaction> map_transaction = MySqlDataStoreUtilities.getTransactions();
				HashMap<String, HashMap<String, Object>> map_daily_sales = new HashMap();

        for(Map.Entry<Integer, Transaction> entry : map_transaction.entrySet()){
          Transaction t = entry.getValue();
          if(t.is_cancelled == false && t.service_status.equalsIgnoreCase("accepted") && t.transaction_status.equalsIgnoreCase("approved")){
            String date = t.booking_date;
            double sales = t.service_total_amount;

            HashMap<String, Object> map_ds = map_daily_sales.getOrDefault(date, new HashMap());
            map_ds.put("date", date);
            map_ds.put("sales", (double)map_ds.getOrDefault("sales", 0.0) + sales);
            map_daily_sales.put(date, map_ds);
          }
        }

				ArrayList<HashMap<String, Object>> list_daily_sales = new ArrayList();
				for(Map.Entry<String,HashMap<String, Object>> entry : map_daily_sales.entrySet()){
					list_daily_sales.add(entry.getValue());
				}
				dataJson = new Gson().toJson(list_daily_sales);
			}

      response.setContentType("application/JSON");
      response.setCharacterEncoding("UTF-8");
      response.getWriter().write(dataJson);

    } catch(Exception e) {
        System.out.println(e.getMessage());
    }
  }

}
